04-onion-routing.md

```diff
@@ -367,6 +379,155 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Route Blinding
+
+Nodes receiving onion packets may hide their identity from senders by
+"blinding" an arbitrary amount of hops at the end of an onion path.
+
+When using route blinding, nodes find a route to themselves from a given
+"introduction node". They then use ECDH with each node in that route to create
+a "blinded" node ID and an encrypted blob (`encrypted_data`) for each one of
+the blinded nodes.
+
+They communicate this blinded route and the encrypted blobs to the sender.
+The sender finds a route to the introduction node and extends it with the
+blinded route provided by the recipient. The sender includes the encrypted
+blobs in the corresponding onion payloads: they allow nodes in the blinded
+part of the route to "unblind" the next node and correctly forward the packet.
+
+The `encrypted_data` is a TLV stream, encrypted for a given blinded node, that
+may contain the following TLV fields:
+
+1. `tlv_stream`: `encrypted_data_tlv`
+2. types:
+    1. type: 1 (`padding`)
+    2. data:
+        * [`...*byte`:`padding`]
+    1. type: 2 (`short_channel_id`)
+    2. data:
+        * [`short_channel_id`:`short_channel_id`]
+    1. type: 4 (`next_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+    1. type: 6 (`path_id`)
+    2. data:
+        * [`...*byte`:`data`]
+    1. type: 8 (`next_blinding_override`)
+    2. data:
+        * [`point`:`blinding`]
+
+#### Requirements
+
+A recipient N(r) creating a blinded route `N(0) -> N(1) -> ... -> N(r)` to itself:
+
+- MUST create a blinded node ID `B(i)` for each node using the following algorithm:
+  - `e(0) <- {0;1}^256`
+  - `E(0) = e(0) * G`
+  - For every node in the route:
+    - let `N(i) = k(i) * G` be the `node_id` (`k(i)` is `N(i)`'s private key)
+    - `ss(i) = SHA256(e(i) * N(i)) = SHA256(k(i) * E(i))` (ECDH shared secret known only by `N(r)` and `N(i)`)
+    - `B(i) = HMAC256("blinded_node_id", ss(i)) * N(i)` (blinded `node_id` for `N(i)`, private key known only by `N(i)`)
+    - `rho(i) = HMAC256("rho", ss(i))`
+    - `e(i+1) = SHA256(E(i) || ss(i)) * e(i)` (ephemeral private key, only known by `N(r)`)
+    - `E(i+1) = SHA256(E(i) || ss(i)) * E(i)` (NB: `N(i)` MUST NOT learn `e(i)`)
+- SHOULD create an `encrypted_data` for each blinded node `B(i)`
+- If it creates `encrypted_data`:
+  - MUST encrypt it with ChaCha20-Poly1305 using the `rho(i)` key and an all-zero nonce
+  - MAY store data in the `path_id` of the payload to itself to verify that
+    the route is used in the right context
+  - MAY add padding data to ensure all `encrypted_data` have the same length
+- MUST communicate the blinded node IDs `B(i)` and `encrypted_data(i)` to the sender
+- MUST communicate the real node ID of the introduction point `N(0)` to the sender
+- MUST communicate the first ephemeral key `E(0)` to the sender
```

Comment:
You're touching on the part that is hard to clarify! I agree this is confusing, but that's because there is a lot of intermediate cryptographic material involved in this blinding scheme. I chose the current naming because I think there are two different layers for this proposal:

- the "low-level" layer where all the key derivations are done, very similar to Sphinx, which really the description of a cryptographic scheme without any interaction with higher-level lightning stuff (messages, tlvs, etc)
- the "high-level" layer where we explain what data should be included in lightning messages

In the "low-level" layer I chose to name things as similarly to Sphinx as possible: this is why `E(i)` is called ephemeral keys, because they're exactly the same as Sphinx's ephemeral keys. At that layer, we also explicitly name the blinding factor `b(i)` because it helps understand where we get some blinding actually done.

In the "high-level" layer, things are named "blinding-something" because they relate to route blinding (we are adding tlvs to messages that contain many other public keys in other tlvs for various use-cases). That is why `E(i)` is actually named blinding point at that layer. However, I don't think this should bleed into the "low-level" layer because at the "low-level" layer, it's somewhat inaccurate (or at least not accurate enough).

Does that make sense?

> This had me thinking the blinded node ID keys were to be included in the onion! Which I'm fairly certain is not being proposed.

I don't understand what you mean, can you clarify?

Reply:
Let me know if I am with you. So in the "high-level", the "blinding" in `blinding_point` signifies that this point is used for the "route blinding scheme" to differentiate it from any additional keys which may be included in the arbitrary TLV data.

In the "low level", the keys *`E(i)`* are similar to Sphinx in that they are ephemeral keys used to establish a secret shared between key generator and routing nodes. It could be imprecise to label them as "blinding" in this context as they are not technically used to blind anything directly. That makes sense to me.

>> This had me thinking the blinded node ID keys were to be included in the onion! Which I'm fairly certain is not being proposed.

> I don't understand what you mean, can you clarify?

I was confused initially and thought that the blinded node ID keys, *`B(i)`* were delivered to each intermediate hop in a blinded route via TLV onion payload using the `blinding_point` field. No worries, it's clear now that is incorrect.

A routing node in a blinded route does not initially know “its” blinded ID public key. This key is selected for the routing node by the blinded route builder and  ***must*** be learned/computed by the routing node ***before*** it can decrypt the onion. So blinded node ID keys cannot be included in onion TLV hop payloads as that would be encrypting the very information needed to decrypt the onion!

Reply:
> Let me know if I am with you. So in the "high-level", the "blinding" in blinding_point signifies that this point is used for the "route blinding scheme" to differentiate it from any additional keys which may be included in the arbitrary TLV data.
>
> In the "low level", the keys E(i) are similar to Sphinx in that they are ephemeral keys used to establish a secret shared between key generator and routing nodes. It could be imprecise to label them as "blinding" in this context as they are not technically used to blind anything directly. That makes sense to me.

That's exactly it! I'm preparing another update to this PR where I'll improve the `proposals` doc, and name `E(i)` blinding ephemeral keys, which is a bit more verbose but more accurate.

> A routing node in a blinded route does not initially know “its” blinded ID public key. This key is selected for the routing node by the blinded route builder and must be learned/computed by the routing node before it can decrypt the onion. So blinded node ID keys cannot be included in onion TLV hop payloads as that would be encrypting the very information needed to decrypt the onion!

That's right! The update to this PR I'm working on should improve the `proposals` doc, which will help make the flow more understandable. However, this scheme is more complex than I hoped it would be, so it will still be quite a tough read ;)